File: packages/verification/src/TraceProof.ts
import { ChdbClient } from "@firegrid/observability"
import * as Effect from "effect/Effect"

import { bindTrialSql, expandTraceMacros } from "./TraceViews.ts"
import { VerificationError } from "./VerificationError.ts"

export interface TraceProof {
  readonly name: string
  readonly sql: string
}

export interface TraceOperationMatch {
  readonly operation: string
  readonly status?: string
  readonly attributes?: Record<string, string | number | boolean>
  readonly outputContains?: ReadonlyArray<string>
  readonly count?: number
}

export interface TraceProofResult {
  readonly ok: boolean
  readonly reason?: string
}

export const truthy = (value: unknown): boolean => {
  if (typeof value === "boolean") return value
  if (typeof value === "number") return value !== 0
  if (typeof value === "bigint") return value !== BigInt(0)
  if (typeof value === "string") return value !== "" && value !== "0" && value.toLowerCase() !== "false"
  return value != null
}

export const normalizeProofSql = (sql: string): string => {
  const trimmed = sql.trim().replace(/;+\s*$/u, "")
  if (!/^(select|with)\b/iu.test(trimmed)) {
    throw new VerificationError({ message: "trace proof SQL must be a SELECT or WITH query" })
  }
  if (trimmed.includes(";")) {
    throw new VerificationError({ message: "trace proof SQL must contain a single read-only query" })
  }
  return expandTraceMacros(trimmed)
}

export const traceSql = (name: string, sql: string): TraceProof => ({
  name,
  sql: normalizeProofSql(sql)
})

const sqlString = (value: string): string => `'${value.replace(/\\/g, "\\\\").replace(/'/g, "\\'")}'`

const spanAttributeEquals = (name: string, value: string | number | boolean): string =>
  `SpanAttributes[${sqlString(name)}] = ${sqlString(String(value))}`

export const traceOperation = (name: string, match: TraceOperationMatch): TraceProof => {
  const conditions = [
    "SpanName = 'verification.operation'",
    spanAttributeEquals("firegrid.operation.name", match.operation),
    ...(match.status === undefined ? [] : [spanAttributeEquals("firegrid.operation.status", match.status)]),
    ...Object.entries(match.attributes ?? {}).map(([attribute, value]) => spanAttributeEquals(attribute, value)),
    ...(match.outputContains ?? []).map((fragment) =>
      `position(SpanAttributes['firegrid.operation.output.json'], ${sqlString(fragment)}) > 0`
    )
  ]
  return traceSql(
    name,
    `
      SELECT countIf(
        ${conditions.join("\n        AND ")}
      ) = ${match.count ?? 1} AS ok
      FROM trial_spans
    `
  )
}

export const runTraceProof = Effect.fn("runTraceProof")(function*(
  proof: TraceProof,
  trialId: string
) {
  const chdb = yield* ChdbClient
  const rows = yield* chdb.unsafe<Record<string, unknown>>(bindTrialSql(proof.sql, trialId)).pipe(
    Effect.mapError((cause) =>
      new VerificationError({
        message: `trace proof ${proof.name} query failed`,
        cause
      })
    )
  )
  const row = rows[0]
  if (row === undefined) {
    return yield* new VerificationError({ message: `trace proof ${proof.name} failed: query returned no rows` })
  }
  if ("ok" in row) {
    if (truthy(row.ok)) return
    return yield* new VerificationError({ message: `trace proof ${proof.name} failed: ok was false` })
  }
  const values = Object.values(row)
  if (values.length === 0) {
    return yield* new VerificationError({ message: `trace proof ${proof.name} failed: query returned an empty row` })
  }
  if (truthy(values[0])) return
  return yield* new VerificationError({ message: `trace proof ${proof.name} failed: first column was false` })
})

File: src/Firegrid.Foundation.Proofs/TraceProof.fs
namespace Firegrid.Foundation.Proofs

open System
open System.Text.RegularExpressions

type TraceProof = { Name: string; Sql: string }

type TraceOperationMatch =
    { Operation: string
      Status: string option
      Attributes: (string * string) list
      OutputContains: string list
      Count: int option }

module TraceOperationMatch =
    let named operation =
        { Operation = operation
          Status = None
          Attributes = []
          OutputContains = []
          Count = None }

    let status value matchSpec = { matchSpec with Status = Some value }

    let attribute key value matchSpec =
        { matchSpec with
            Attributes = matchSpec.Attributes @ [ key, value ] }

    let outputContains value matchSpec =
        { matchSpec with
            OutputContains = matchSpec.OutputContains @ [ value ] }

    let exactly count matchSpec = { matchSpec with Count = Some count }

module TraceProof =
    let private trialSpansSql =
        """
(
  SELECT *
  FROM file({spans_jsonl:String}, JSONEachRow)
  WHERE trial_id = {trial_id:String}
)
"""

    let private verificationOperationsSql =
        """
(
  SELECT
    trial_id,
    JSONExtractString(toJSONString(attributes), 'firegrid.client.id') AS client_id,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.id') AS operation_id,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.name') AS operation,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.key') AS operation_key,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.input.json') AS input_json,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.output.json') AS output_json,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.status') AS status,
    JSONExtractString(toJSONString(attributes), 'firegrid.operation.failure.kind') AS failure_kind
  FROM trial_spans
  WHERE name = 'verification.operation'
)
"""

    let private hasTraceMacro (sql: string) =
        Regex.IsMatch(sql, @"\b(trial_spans|verification_operations)\b", RegexOptions.IgnoreCase)

    let private containsForbiddenReader (sql: string) =
        Regex.IsMatch(sql, @"\b(file|url|s3|hdfs|mysql|postgresql|remote|cluster)\s*\(", RegexOptions.IgnoreCase)
        || Regex.IsMatch(sql, @"\bsystem\s*\.", RegexOptions.IgnoreCase)

    let private expandTraceMacros (sql: string) =
        sql
        |> fun value ->
            Regex.Replace(value, @"\bverification_operations\b", verificationOperationsSql, RegexOptions.IgnoreCase)
        |> fun value -> Regex.Replace(value, @"\btrial_spans\b", trialSpansSql, RegexOptions.IgnoreCase)

    let normalizeSql (sql: string) =
        let trimmed = sql.Trim().TrimEnd(';').Trim()

        if
            not (
                trimmed.StartsWith("select", StringComparison.OrdinalIgnoreCase)
                || trimmed.StartsWith("with", StringComparison.OrdinalIgnoreCase)
            )
        then
            failwith "trace proof SQL must be a SELECT or WITH query"

        if trimmed.Contains(";") then
            failwith "trace proof SQL must contain one read-only query"

        if not (hasTraceMacro trimmed) then
            failwith "trace proof SQL must query trial_spans or verification_operations"

        if containsForbiddenReader trimmed then
            failwith "trace proof SQL cannot use external table readers"

        expandTraceMacros trimmed

    let sql name body =
        { Name = name; Sql = normalizeSql body }

    let private sqlString (value: string) =
        "'" + value.Replace("\\", "\\\\").Replace("'", "\\'") + "'"

    let private attributeField (key: string) =
        "JSONExtractString(toJSONString(attributes), " + sqlString key + ")"

    let private attributeEquals key value =
        sprintf "%s = %s" (attributeField key) (sqlString value)

    let operation name matchSpec =
        let conditions =
            [ "name = 'verification.operation'"
              attributeEquals "firegrid.operation.name" matchSpec.Operation
              match matchSpec.Status with
              | Some status -> attributeEquals "firegrid.operation.status" status
              | None -> ""
              yield! matchSpec.Attributes |> List.map (fun (key, value) -> attributeEquals key value)
              yield!
                  matchSpec.OutputContains
                  |> List.map (fun value ->
                      sprintf
                          "position(%s, %s) > 0"
                          (attributeField "firegrid.operation.output.json")
                          (sqlString value)) ]
            |> List.filter (fun condition -> condition <> "")

        let expectedCount = matchSpec.Count |> Option.defaultValue 1

        sql
            name
            (sprintf
                """
SELECT countIf(
  %s
) = %d AS ok
FROM trial_spans
"""
                (String.concat "\n  AND " conditions)
                expectedCount)

    let asCheck (proof: TraceProof) : Check<'result> =
        { Name = proof.Name
          RunCheck =
            fun trial ->
                async {
                    let! result =
                        TraceSql.scalarTruthy
                            { TrialId = trial.TrialId
                              SpansJsonl = trial.Traces.SpansJsonl }
                            { Sql = proof.Sql
                              Parameters = Map.empty }

                    return result |> Result.mapError (fun message -> "trace proof failed: " + message)
                } }
